### ISSUE TITLE
Controller for Nav

### ISSUE THREAD

**Issue by zeroender** [susanBuck/dwa15-fall2017#220]
My navigation section which contains some PHP and html has expanded to the point where I'm pretty sure it violates the seperation of display and logic code (the nav has several dropdowns, dynamically populated with data from an eloquent query).  Is it appropriate to create a seperate "NavController" and "NavView", and include them that way?

**Comment by davida26** [comment]
could you elaborate a bit more on what logic you are adding here?
Is this nav specific to one route or shared across?

In my opinion if your nav is requiring heavy use of logic, then consider whether the nav is becoming too complex for the user.

Also dont forget that within your specific controllers you can pass variables to the nav. 

**Example:**
- you have a controller for animals, your index shows a list of animals.
- You want the nav to output a variable when the user is at the animals/ route (index).
- you can pass in your animals controller a variable (manipulate some logic before passing it within the animal controller) 

In your AnimalController.php

```php
public function index() {
$test = 'hello world';
# some logic performed here on $test if you wanted to
return view('animals.index')->with(['test' => $test]);
}
```

then in your nav you can check to see if that isset:

```php
@if(isset($test)) {{ $test }}  @endif
```
 
So now in the animals.index i can see "hello world" but anywhere else I will not. 

In other controllers you can add that variable again and pass it to the view. With different text in this case or logic in your case to output something different.

Depends on your use case of course.

**Comment by susanBuck** [comment]
Good notes from @davida26 above. 
I also recommend sharing your nav code with us so we can discuss possible ways to refactor.

**Comment by zeroender** [comment]
That's a very interesting case - it's not quite what I'm doing, but my current code could be adapted somewhat to do that.

My full code is below, here's a quick summary of what it is doing:

The nav header has a number of dropdowns, the first allows a user to create tasks or categories (my two models).  The dropdown here is static.  The next dropdown takes the user to an index for the tasks, and has several links.  each of these call the index, passing a different variable to an eloquent request, resulting in an index page sorted by various attributes.  Next is a dropdown that contains an index and dynamically populated list (by a database query) of every category.  Finally is a dropdown that allows the user to choose between various queries done to return tasks based on the category they are attached too.

So... I have a lot of logic going on in the nav!  

@php

    use App\Category; 
    $categories = Category::orderBy('name')->get();
    
    $categoryNav = [
        'category' => 'All Categories',
        'SECTIONBREAK' => 'SECTIONBREAK'
    ];
    
    foreach ($categories as $category) {
        $categoryNav+= array('category/'.$category->id => $category->name);
    }

    $taskByCategoryNav = [];
    foreach ($categories as $category) {
        $taskByCategoryNav+= array('category/'.$category->id.'/tasks' => $category->name);
    }

    $taskNav = [
        'task' => 'View all Tasks',
        'SECTIONBREAK' => 'SECTIONBREAK',
        'taskBy/name' => 'View by Name',
        'taskBy/description' => 'View by Description',
        'taskBy/due_date' => 'View by Date',
        'taskBy/status' => 'View by Status'
    ];

    $createNav = [
        'task/create' => 'Create a new task',
        'category/create' => 'Create a new category'
    ]
@endphp

<nav class="navbar navbar-expand-lg navbar-light bg-light">
  <div class="collapse navbar-collapse" id="navbarSupportedContent">
    <ul class="navbar-nav mr-auto">

      <li class="nav-item dropdown">
        <a class="nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
          Create
        </a>
        <div class="dropdown-menu" aria-labelledby="Tasks">
          @foreach($createNav as $link => $label)
            @if($link != 'SECTIONBREAK')
              <a class="dropdown-item" href='/{{ $link }}'>{{ $label }}</a>
            @else
              <div class="dropdown-divider"></div>
            @endif
          @endforeach
        </div>
      </li>

      <li class="nav-item dropdown">
        <a class="nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
          Tasks
        </a>
        <div class="dropdown-menu" aria-labelledby="Tasks">
          @foreach($taskNav as $link => $label)
            @if($link != 'SECTIONBREAK')
              <a class="dropdown-item" href='/{{ $link }}'>{{ $label }}</a>
            @else
              <div class="dropdown-divider"></div>
            @endif
          @endforeach
        </div>
      </li>

      <li class="nav-item dropdown">
        <a class="nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
          Categories
        </a>
        <div class="dropdown-menu" aria-labelledby="Tasks">
          @foreach($categoryNav as $link => $label)
            @if($link == 'SECTIONBREAK')
              <div class="dropdown-divider"></div>
            @else
              <a class="dropdown-item" href='/{{ $link }}'>{{ $label }}</a>
            @endif
          @endforeach
        </div>
      </li>

      <li class="nav-item dropdown">
        <a class="nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
          Tasks by Category
        </a>
        <div class="dropdown-menu" aria-labelledby="Tasks">
          @foreach($taskByCategoryNav as $link => $label)
            @if($link == 'SECTIONBREAK')
              <div class="dropdown-divider"></div>
            @else
              <a class="dropdown-item" href='/{{ $link }}'>{{ $label }}</a>
            @endif
          @endforeach
        </div>
      </li>

    </ul>
  </div>
</nav>

**Comment by zeroender** [comment]
Going through all issues now and I think my question might be answered https://github.com/susanBuck/dwa15-fall2017/issues/207

**Comment by zeroender** [comment]
Update - I've added a ViewServiceProvider just for the nav, and was able to move all the php into that file, which serves just the nav view.  Works great!

Two follow up questions:

Is it bad to have lots of static php code in ViewServiceProvider?  Pretty much all of my PHP ended up there.

Is there a way to get the variable(s) from the ViewServiceProvider into the Controller, rather than a view?  I was able to get the database request I do under ViewServiceProvider (a general request for all categories) to be available on every view, but it does the actual request with the load of each view.  Since the nav is always present, there are a lot of scenarios where the same data is loaded twice.  Not terrible, but could be more efficient if I had the actual object instead and could pass that instead (and use methods on that data to get specific rather than hitting the database again).

**Comment by susanBuck** [comment]
That was quite a bit of PHP/logic in your view, so it'd good that you've refactored it out, and I think ViewServiceProviders are perfect for this.

**re: Q1:**
I don't think it's bad to have the static PHP code in the ViewServiceProvider.

**re: Q2:**
Ah yes, the use of the wildcard symbol for view composers does have some inefficiencies.

To demonstrate this, I added a dump statement:

```php
public function boot()
    {
        view()->composer('*', function ($view) {
            dump('Invoked the ViewServiceProvider'); # <---
            $user = request()->user();
            $view->with('user', $user);
        });
    }
```

And you can see that for a single page, we have that same code being run **five times**:

<img width="974" alt="image" src="https://user-images.githubusercontent.com/237711/33899306-14dff62c-df39-11e7-955d-3f5b57f6bcee.png">

To optimize this, I made it so the composer is available to just `layout.master`, which @includes `modules.nav`, so the data is available there as well:

```php
public function boot()
    {
        view()->composer('layouts.master', function ($view) {
            $user = request()->user();
            $view->with('user', $user);
        });
    }
```
This reduces the invocation down to x1 time.

(Of course, if I wanted the user data on any of the specific page views, I'd have to update the composer to include those views, or just pass the data when that view is returned.)

Does the above refactoring address your repetitive database calls concerns?

**Comment by zeroender** [comment]
That pretty much covers it, thanks!

I found that including that data just on the nav had the same affect - the queries were run just once for the nav, and in some situations were run again if the current page was hitting the same query.  I did try sharing the data to other views, which does make it available but each view preforms a seperate request - I don't think with the ViewServiceProvider there is a way around that (if the same request occurs in two places).
